### PR TITLE
Health Checker major updates

### LINKED ISSUE #546: HealthChecker - Add Ability to pass multiple servers
**Describe The Request**
Need to add a parameter called -Servers that would be able to execute the script on each server in the list all at once. This would allow the script to complete quicker. 

Migrated From: https://github.com/dpaulson45/HealthChecker/issues/158

### LINKED ISSUE #1321: [Issue] - Health Checker display of CVE-2022-21978
**Provide Version Number**
22.10.28.0217

**Describe the issue**
Security Information has the below string

Security Vulnerability: CVE-2022-21978
		**Install the May 2022 SU** and run /PrepareDomain or /PrepareAllDomains - See: https://aka.ms/HC-May22SU

**Expected behavior**
Security Vulnerability: CVE-2022-21978
		**Install any SU greater than May 2022 SU** and run /PrepareDomain or /PrepareAllDomains - See: https://aka.ms/HC-May22SU

**Script Output**
If applicable, add the exception that wasn't handled.

**Additional context**

### PR DESCRIPTION
**Pull request by dpaulson45**
**Reason:**
Changed up to allow Health Checker to pass a list of servers, including from the pipeline. In order to be more efficient, wanted to avoid making calls that aren't going to change from each server. 

Still need to verify testing from Edge and how to handle from a 2013 server against a 2019 box. This would be ideal before merge and release. However, this will not be completed till after my return from vacation. 

**Fix:**
Short description of the fix

**Validation:**

Resolved #1321 
Resolved #546

### CODE REVIEW

**Comment by lusassl-msft** [comment]
Just for tracking purposes:

Track the split permissions indicator by the help of the custom class and improve the check to avoid using `throw` in non-error scenarios: https://github.com/microsoft/CSS-Exchange/pull/1350

File: Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1 [hunk 1]
```diff
@@ -72,19 +71,18 @@ function Invoke-AnalyzerExchangeInformation {
...
+    $extendedSupportDate = $exchangeInformation.BuildInformation.VersionInformation.ExtendedSupportDate
     if ($extendedSupportDate -le ([DateTime]::Now.AddYears(1))) {
         $displayWriteType = "Yellow"
 
         if ($extendedSupportDate -le ([DateTime]::Now.AddDays(178))) {
             $displayWriteType = "Red"
         }
 
-        $displayValue = "$($exchangeInformation.BuildInformation.ExtendedSupportDate) - Please note of the End Of Life date and plan to migrate soon."
+        $displayValue = "$($exchangeInformation.BuildInformation.VersionInformation.ExtendedSupportDate.ToString("MM/dd/yyyy")) - Please note of the End Of Life date and plan to migrate soon."
```
Comment (lusassl-msft) [inline Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1:82]: I'm wondering if it makes more sense to display the support date like this: `Tuesday, April 11, 2023`.
Would be easier to understand as the en-US notation might confuse people who normally use the `dd/MM/yyyy` notation.

File: Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1 [hunk 2]
```diff
@@ -72,19 +71,18 @@ function Invoke-AnalyzerExchangeInformation {
...
 
         if ($extendedSupportDate -le ([DateTime]::Now)) {
-            $displayValue = "$($exchangeInformation.BuildInformation.ExtendedSupportDate) - Error: You are past the End Of Life of Exchange."
+            $displayValue = "$($exchangeInformation.BuildInformation.VersionInformation.ExtendedSupportDate.ToString("MM/dd/yyyy")) - Error: You are past the End Of Life of Exchange."
```
Comment (lusassl-msft) [inline Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1:85]: Same here. I think `Tuesday, April 11, 2023` is better than `MM/dd/yyyy`

File: Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1 (hunk 1 above)
Comment (dpaulson45) [reply Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1:82]: @lusassl-msft does it convert the name of the month if the OS is not English? 

```
$date = get-date
$date.ToString("MMM dd, yyyy")
$date.ToString("MMMM dd, yyyy")
```

If it does, we would also want to force it to be English so when it writes out to the file it should be able to be read by anyone.

Comment (lusassl-msft) [reply Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1:82]: Tested on German OS:
image

Comment (lusassl-msft) [reply Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1:82]: Returns: `Dec 08, 2022`

Comment (dpaulson45) [reply Diagnostics/HealthChecker/Analyzer/Invoke-AnalyzerExchangeInformation.ps1:82]: Looks like those are the only two places we display in that format.

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeInformation.ps1 [hunk 3]
```diff
@@ -409,94 +151,55 @@ function Get-ExchangeInformation {
...
-            } catch {
-                Write-Yellow "Failed to run Get-HybridConfiguration"
-                Invoke-CatchActions
-            }
+            $exchangeInformation.ApplicationPools = Get-ExchangeAppPoolsInformation -Server $Server
 
             Write-Verbose "Query Exchange Connector settings via 'Get-ExchangeConnectors'"
             $exchangeInformation.ExchangeConnectors = Get-ExchangeConnectors `
-                -ComputerName $Script:Server `
+                -ComputerName $Server `
```
Comment (lusassl-msft) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeInformation.ps1:178]: Here as well?

File: Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeUpdates.ps1 [hunk 4]
```diff
@@ -17,7 +21,7 @@ function Get-ExchangeUpdates {
         $RegLocation = "SOFTWARE\Microsoft\Updates\Exchange 2019"
     }
 
-    $RegKey = Get-RemoteRegistrySubKey -MachineName $Script:Server `
+    $RegKey = Get-RemoteRegistrySubKey -MachineName $Server `
```
Comment (lusassl-msft) [inline Diagnostics/HealthChecker/DataCollection/ExchangeInformation/Get-ExchangeUpdates.ps1:24]: Not a hard requirement but as we do these many changes, we could do splatting here too.

File: Diagnostics/HealthChecker/DataCollection/ServerInformation/Get-HttpProxySetting.ps1 [hunk 5]
```diff
@@ -39,13 +43,13 @@ function Get-HttpProxySetting {
         }
     }
 
-    $httpProxy32 = Invoke-ScriptBlockHandler -ComputerName $Script:Server `
+    $httpProxy32 = Invoke-ScriptBlockHandler -ComputerName $Server `
```
Comment (lusassl-msft) [inline Diagnostics/HealthChecker/DataCollection/ServerInformation/Get-HttpProxySetting.ps1:46]: Do splatting would be good here as well

File: Diagnostics/HealthChecker/Helpers/Invoke-SetOutputInstanceLocation.ps1 [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        [Parameter(Mandatory = $false)]
+        [bool]$IncludeServerName = $false
+    )
+    $endName = "-{0}.txt" -f $Script:dateTimeStringFormat
+
+    if ($IncludeServerName) {
+        $endName = "-{0}{1}" -f $Server, $endName
+    }
+
+    $Script:OutputFullPath = "{0}\{1}{2}" -f $Script:OutputFilePath, $FileName, $endName
```
Comment (lusassl-msft) [inline Diagnostics/HealthChecker/Helpers/Invoke-SetOutputInstanceLocation.ps1:21]: I think this one should also work
```suggestion
    $Script:OutputFullPath = [System.IO.Path]::Combine($Script:OutputFilePath, $FileName, $endName)
```

**Review by lusassl-msft — APPROVED** [review]
Only minor finding. Lab testing runs also smoothly. Approved.

File: Diagnostics/HealthChecker/Helpers/Invoke-SetOutputInstanceLocation.ps1 (hunk 6 above)
Comment (dpaulson45) [reply Diagnostics/HealthChecker/Helpers/Invoke-SetOutputInstanceLocation.ps1:21]: This actually wouldn't work as you have it.

image

But could do something like this: 

image

Comment (lusassl-msft) [reply Diagnostics/HealthChecker/Helpers/Invoke-SetOutputInstanceLocation.ps1:21]: You're right. Let's use your suggestion or keep it as is.
